### PR TITLE
Let swarmkit to handle cluster defaults in `swarm init` if not specified

### LINKED ISSUE #24958: Cluster defaults should not be overridden by swarm init
`swarm init` creates a `ClusterSpec` and overwrites the cluster's object spec with this one. This means defaults such as certificate validity period and task history retention are under the client's control. They should be under SwarmKit's control instead.

`swarm init` should only change values from the defaults when a value is given through a flag. Currently, it appears to overwrite the whole spec with the client-side defaults when no flags are specified.

cc @aluzzardi @tonistiigi

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to address the issue raised in #24958 where previously `docker swarm init` will automatically fill in all the default value (instead of letting swarmkit to handle the default).

**\- How I did it**

This fix update the `swarm init` so that initial value are passed only when a flag change has been detected.

In addition, ClusterSpec in engine-api has been updated to use pointers so that unchanged values could be detected.

**\- How to verify it**

Existing tests passed. In addition, manual tests has been done to verify with `1.13.0 client vs. 1.13.0 daemon` and `1.12.0 client vs. 1.13.0 daemon`

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

A pull request in engine-api will be created separately.

This fix fixes #24958.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

**Comment by aaronlehmann** [comment]
Thanks for working on this!

I'm trying to decide if changing all fields under `Cluster` to pointers is the right approach. The risk is that we'll end up with something like the AWS SDK where everything is a pointer and this causes a lot of awkwardness. But I'm not sure I have a better idea. Maybe there should be a seperate version of `Cluster` just for making changes, so the use cases that retrieve the information aren't affected - but I'm not sure this is better. It will be important to get this right, because we'll have to make a similar change to `Service` for #24959.

ping @docker/core-engine-api-maintainers @stevvooe - any thoughts?

**Comment by stevvooe** [comment]
I really don't like having pointers everywhere. It makes using the fields challenging and destroys performance for fields with a high access frequency. Everything possible should be done to make the zero value useful or resolve defaults at the application edge. I think we've already done this for the most part.

The main problem with the pattern of pointers is that we are pushing the knowledge about whether or not a field was set into the backend when it is really just a UI problem. Avoid requiring knowledge about setting deep in the backend (a huge mistake made in AWS) and this isn't a problem.

Well formed code should allow something as simple as this:

```go
var config Config
config = *defaultConfig

// now, set the field from the command line
if flag.Changed("foo") {
  config.SetFoo(flag.Value("foo"))
}
```

The better way of dealing with this is to get the original, touch only the modified fields and send back the result. We do this for most other API objects, but I don't think join/leave got that careful treatment:

```
config := getCurrent()

if flag.Changed("foo") {
   config.SetFoo(flag.Value("foo"))
}

updateConfig(config)
```

**Comment by aaronlehmann** [comment]
This case is tricky because it deals with setting the initial values, not doing an update. `TaskHistoryRetentionLimit` is a good example. 0 is a valid value, but not the default. However, we want to leave the defaults up to swarmkit instead of having the engine insert its own default.

**Comment by aaronlehmann** [comment]
Design LGTM unless someone has a better idea.

**Comment by stevvooe** [comment]
> Design LGTM unless someone has a better idea.

> __Everything possible should be done to make the zero value useful or resolve defaults at the application edge.__

**Comment by stevvooe** [comment]
If that wasn't clear: NOT LGTM.

Please make the zero-value useful or redesign the data structure to do it.

Making everything pointers is the lazy way out here. Most of these fields aren't realistic with the zero-value.

**Comment by aaronlehmann** [comment]
The zero value is not useful here. I explained that above: https://github.com/docker/docker/pull/25697#issuecomment-239969944

I know you have good instincts about this kind of thing (which is why I pinged you), and I would really appreciate your help finding a good solution. But as it is, I don't understand what you think should be done.

File: vendor/src/github.com/docker/engine-api/types/swarm/swarm.go
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/swarm/swarm.go]: This should not be a pointer type.

**Comment by stevvooe** [comment]
For `TaskHistoryRetentionLimit`, zero may be a valid value but isn't realistic. We really need to always keep one or the update algorithms break.

More fields:

`SnapshotInterval` - 0 not valid
`KeepOldSnapshots` - 0 maybe valid, but not realistic, always should have 1 snapshot
`LogEntriesForSlowFollowers` - 0 not realistic
`HeartbeatTick` - 0 invalid
`ElectionTick` - 0 invalid

I am just not seeing this as a problem.

Worst case, we export the defaults and do this:

```
var cluster agent.Cluster
cluster = agent.DefaultCluster

// ... then set fields
```

**Comment by aaronlehmann** [comment]
> For TaskHistoryRetentionLimit, zero may be a valid value but isn't realistic. We really need to always keep one or the update algorithms break.

For this value:

- `< 0` means unlimited
- 0 means to delete tasks objects the moment they stop running
- `> 0` means to keep that number of historic tasks. i.e. 1 means to keep only the most recent dead task in history.

Without a 0 value, there is no way to turn off task history. Someone asked how to do this yesterday.

> KeepOldSnapshots - 0 maybe valid, but not realistic, always should have 1 snapshot

0 is the default value. There's no need to keep any *old* snapshots around unless you are being conservative and want them as checkpoints of older state.

For the others, I don't think the 0 value is useful. So perhaps only these two should become pointers, and the rest should not.

I think the `agent.DefaultCluster` approach is not my first choice because it makes certain default behaviors dependent on the CLI version, but maybe that's not the worst thing.

**Comment by stevvooe** [comment]
> 0 means to delete tasks objects the moment they stop running

I think we've bound two concepts together that shouldn't. The user was interested in cleaning up the container resources, which we control with the assignments (I also provided that user with a methodology for cleaning them up _as soon as_ they exit). The fact that his is bound to the history is an artifact of the design. One may well want to keep history but cleanup containers earlier or vice versa.

We should probably have another way of controlling assignment set.

> 0 is the default value. There's no need to keep any old snapshots around unless you are being conservative and want them as checkpoints of older state.

If zero is the default value, then we don't need to do anything here.

**Comment by aaronlehmann** [comment]
@yongtang: Would you mind updating the PR to make `TaskHistoryRetentionLimit` a pointer and use zero-values for the other items to signal defaults?

**Comment by yongtang** [comment]
@aaronlehmann Sure. Let me update the pull request.

**Comment by yongtang** [comment]
@aaronlehmann The pull request has been updated. Please let me know if there are any issues.

File: daemon/cluster/convert/swarm.go
Comment (aaronlehmann) [inline daemon/cluster/convert/swarm.go]: I think it's slightly more accurate to say that if the value is nil or 0 the value won't be changed (maybe it was changed earlier, and it's not the default anymore).

File: daemon/cluster/cluster.go
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: Doesn't this have to pass the spec from `getSwarm` so that defaults aren't overwritten?

Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks @aaronlehmann. At the client side (`api/client/swarm/update.go`) the existing swarm setting was obtained and then merged with the setting updates. So the updated setting is reflected in `spec` here (without `getSwarm`).

But it is better to  obtain the existing swarm setting in daemon side here (with `getSwarm` and then update as needed) as end user may not use client directly (vs. remote API). Processing at the daemon side will prevent incorrect settings through remote API.

Let me update the pull request to change it.

File: daemon/cluster/convert/swarm.go
Comment (yongtang) [inline daemon/cluster/convert/swarm.go]: Thanks @aaronlehmann the pull request has been updated.

File: daemon/cluster/cluster.go
Comment (yongtang) [inline daemon/cluster/cluster.go]: @aaronlehmann The pull request has been updated to address the above mentioned. Please let me know if there are any issues.

Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: > Thanks @aaronlehmann. At the client side (api/client/swarm/update.go) the existing swarm setting was obtained and then merged with the setting updates. So the updated setting is reflected in spec here (without getSwarm).

That makes sense.

> But it is better to obtain the existing swarm setting in daemon side here (with getSwarm and then update as needed) as end user may not use client directly (vs. remote API). Processing at the daemon side will prevent incorrect settings through remote API.

I'm actually now thinking that it's better not to do this at the API level for `swarm update`. If a user sets a value to 0 through the API, they would expect it to get reset to the default, but what happens instead is that the value stays unchanged. We don't have a way to fetch manager-side default values except when we are creating the cluster, so it's difficult to support resetting fields to the defaults.

So I think it's probably better to require a complete spec for `swarm update`, and return an error if anything that shouldn't be `0`/`nil` is set to `0` or `nil`. This also keeps engine-api more consistent with the GRPC api, where the client is expected to always form the spec on its own.

Sorry for the misdirection - it just took me a few passes to wrap my mind around this.

File: api/client/system/info.go
Comment (aaronlehmann) [inline api/client/system/info.go]: To be safe, make sure this is not `nil` before dereferencing it.

File: daemon/cluster/cluster.go
Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks @aaronlehmann. The pull request has been updated. Please let me know if there are any issues.

File: api/client/system/info.go
Comment (yongtang) [inline api/client/system/info.go]: Thanks. The PR has been updated.

**Comment by aaronlehmann** [comment]
Design LGTM

**Comment by cpuguy83** [comment]
Needs a rebase, and engine-api is moved to `github.com/docker/docker/api/types`

**Comment by yongtang** [comment]
Thanks @cpuguy83. The pull request has been updated. And the PR is not depending on engine-api anymore.

**Comment by aaronlehmann** [comment]
LGTM

Would love to get this in for 1.13.

Removing the `status/needs-vendoring` label because this does no longer needs any vendoring.

**Comment by aaronlehmann** [comment]
Sorry, needs another rebase.

If you revendor swarmkit today, I suggest vendoring revision `b79d41fa99c137181d8f58ef76a6e8a25bc2e72f` to avoid conflicts with your other PR, and because swarmkit master has a bad regression at the moment.

And ping for review.

**Comment by thaJeztah** [comment]
Before:

```JSON
{
  "AdvertiseAddr": "",
  "ForceNewCluster": false,
  "ListenAddr": "0.0.0.0:2377",
  "Spec": {
    "CAConfig": {
      "NodeCertExpiry": 7.776e+15
    },
    "Dispatcher": {
      "HeartbeatPeriod": 5000000000.0
    },
    "Orchestration": {
      "TaskHistoryRetentionLimit": 5
    },
    "Raft": {
      "ElectionTick": 0,
      "HeartbeatTick": 0
    },
    "TaskDefaults": {}
  }
}
```

After:

```JSON
{
  "AdvertiseAddr": "",
  "ForceNewCluster": false,
  "ListenAddr": "0.0.0.0:2377",
  "Spec": {
    "CAConfig": {},
    "Dispatcher": {},
    "Orchestration": {},
    "Raft": {
      "ElectionTick": 0,
      "HeartbeatTick": 0
    },
    "TaskDefaults": {}
  }
}
```


Interestingly (but not related to this PR); on docker 1.12.1, `"Raft"` is empty, but on current master `"ElectionTick"`, and `"HeartbeatTick"` are always set.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by yongtang** [comment]
Thanks all for the review and help!

**Pull request merged**
